### ISSUE TITLE
Add `layer-name-ident-allowed-list`

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#8480]
### What is the problem you're trying to solve?

I want to enforce specific strings for the idents within layer names.

For example, only allow "foo", "bar" and "baz" and nothing else.

```css
@layer foo {};
@layer bar;
@layer foo.bar {}
@import "foo.css" layer(baz);
```

See https://github.com/stylelint/stylelint/pull/8474#discussion_r1997527938

### What solution would you like to see?

A new rule:

- Name: `layer-name-ident-allowed-list`
- Description: Specify a list of allowed layer name idents.
- Message: `Unexpected ident "${ident}"`
- Primary option: `array[string|regex]|string|regex`
- Secondary options: None
- Category: Enforce conventions > Allowed, disallowed & required
- Autofix?: No
- Included in sharable configs?: No

I've gone with "ident" over "segment" as segments can be anonymous ([spec](https://drafts.csswg.org/css-cascade-5/#layer-names)):

> A [cascade layer](https://drafts.csswg.org/css-cascade-5/#cascade-layers) has a layer name, which is an ordered list representing each level of layer nesting, each segment of which can be named (as a [CSS identifier](https://drafts.csswg.org/css-values-4/#css-css-identifier)) or anonymous

**Comment by ybiquitous** [comment]
Sorry, I'm not sure about the difference from `layer-name-pattern` (#8431). Can you explain it?

**Comment by ybiquitous** [comment]
Ah, I missed <https://github.com/stylelint/stylelint/pull/8474#discussion_r1997527938> for the background. But I'm still wondering if users might be confused about the two rules. 🤔 

Also, are the rules conflicting by the following rule criteria?

https://github.com/stylelint/stylelint/blob/58020223db8fdcdf61fbc3bf473bd9c2c4a310b9/docs/developer-guide/rules.md?plain=1#L18-L19

**Comment by jeddy3** [comment]
> But I'm still wondering if users might be confused about the two rules. 🤔

It's a fair concern.

Adding this rule will be consistent with our previous approach of adding overlapping rules with slightly different purposes to provide people with a toolkit of configurable rules for various use cases. I don't believe anyone has raised a concern about confusion yet.

For example, the [`pattern`](https://stylelint.io/user-guide/rules#pattern) rules are for general naming patterns whereas the [`list` and `no`](https://stylelint.io/user-guide/rules#allowed-disallowed--required) rules are for (dis)allowing specific things, and the [`max`](https://stylelint.io/user-guide/rules#max--min) for limiting things.

A specific example would be... if someone wants to:

- disallow all id selectors they can use [`selector-max-id`](https://stylelint.io/user-guide/rules/selector-max-id)
- disallow a specific id selector they can use [`selector-disallowed-list`](https://stylelint.io/user-guide/rules/selector-disallowed-list)
- allow id selectors but ensure they're lowercase, they can use [`selector-id-pattern`](https://stylelint.io/user-guide/rules/selector-id-pattern)

Arguably, all three use cases could be covered by `selector-disallowed-list` because it accepts regex, but it becomes unwieldy to configure, so we offer a toolkit of rules instead that often accept strings. 

When it comes to layer names, these new rules will fall into these preexisting categories:

- `layer-name-pattern` for people not wanting to enforce specific names but wanting the names they do use to follow a convention like lowercase
- `layer-name-ident-allowed-list` for people wanting to enforce specific names

For example:

```json
{
  "rules": {
    "layer-name-pattern": "^[a-z][a-z0-9.-]*$",
  }
}
```

Or:

```json
{
  "rules": {
    "layer-name-ident-allowed-list": ["defaults", "components", "theme"],
  }
}
```

Admittedly, this ease of use introduces some redundancy, and there's merit to both approaches.

I'm optimistic that users won't find it confusing because the rules will be consistent with those we offer for other constructs, like selectors and declarations.

**Comment by ybiquitous** [comment]
Thanks for the explanation. I understand the advantage of separating the rules.

However, what do you think about the following examples?

```json
{
  "rules": {
    "layer-name-pattern": "^(defaults|components|theme)$"
  }
}
```

```json
{
  "rules": {
    "layer-name-ident-allowed-list": ["defaults", "components", "theme"]
  }
}
```

Besides, what about nested layers?

```css
/* Only the following patterns are allowed. */
@layer framework.base {}
@layer framework.theme {}
```

```json
{
  "rules": {
    "layer-name-pattern": "^framework\\.(base|theme)$"
  }
}
```


```jsonc
{
  "rules": {
    "layer-name-ident-allowed-list": ["framework", "base"]
    //=> Invalid patterns are allowed, e.g. "base.framework"
  }
}
```

Also, what about this case?

```css
/* Only the following patterns are allowed. */
@layer framework.base {}
@layer framework.theme {}
@layer framework {
  @layer base {}
}
@layer framework {
  @layer theme {}
}
```

```json
{
  "rules": {
    "layer-name-pattern": "^(framework\\.(base|theme)|framework|base)$"
  }
}
```


```jsonc
{
  "rules": {
    "layer-name-ident-allowed-list": ["framework", "base"]
    //=> Invalid patterns are allowed, e.g. "base.framework"
  }
}
```

In my feelings, `layer-name-ident-allowed-list` works for quite simple cases, while `layer-name-pattern` satisfies most cases. Still, considering `@layer` only, providing `layer-name-pattern` only seems enough to me. 🤔

**Comment by jeddy3** [comment]
All good points. 

> Still, considering `@layer` only, providing `layer-name-pattern` only seems enough to me

I think we can go with this assumption, especially given the unique `.` segments.

Let's leave this issue as a discussion in case anyone else wants to chime in with their use case, though.

**Comment by ybiquitous** [comment]
Yes, I'd like to hear other thoughts. 😃

**Comment by Mouvedia** [comment]
> Let's leave this issue as a discussion in case anyone else wants to chime in with their use case, though.

It's been several months since then, I am closing.
This may be reopened if someone eventually comes with better arguments.
